Introduce mux control API

decsny opened the pull request:
Introduce a way to control muxes based on DT description through a generic API. 
The DT description is based on Linux: https://www.kernel.org/doc/Documentation/devicetree/bindings/mux/mux-controller.txt 
The API refer to Linux: https://github.com/torvalds/linux/blob/master/include/linux/mux/consumer.h

### Overview

This PR introduces a generic MUX (multiplexer) device model for Zephyr and adds
initial controller drivers, devicetree bindings, driver integrations, documentation

It includes:

- generic devicetree bindings for MUX controllers and consumers
- a common MUX driver API
- generic and SoC-specific MUX controller drivers
- helper macros for devicetree-based MUX descriptor definition
- integration of existing drivers with the MUX device model
- peripheral documentation

### Motivation

Signal routing and multiplexer configuration are commonly needed for both SoC
internal routing blocks and external multiplexers. Before this series, support
for this kind of hardware was controller-specific and not expressed through a
common device mode;.

This series addresses the following problems:

- no common API for configuring and querying MUX controllers
- no generic devicetree model for MUX consumers
- signal routing dependencies hidden inside controller-specific or driver-specific code
- duplicated boilerplate when drivers need to define and iterate over multiple
  MUX entries from devicetree

This PR introduces a reusable MUX device model so drivers can describe routing in
devicetree and configure it through a common interface.

### Design Details

This PR adds the following building blocks.

#### 1. Devicetree bindings

Add generic MUX bindings:

- `mux-controller.yaml`
- `mux-consumer.yaml`

Add controller-specific bindings:

- `gpio-mux.yaml`
- `nxp,inputmux.yaml`
- `nxp,trgmux.yaml`
- `nxp,mcux-xbar.yaml`

A consumer node references a mux controller via one of two phandle-array properties:

- mux-controls = <&ctrl addressing-cells> — consumer holds a reference to a control line and supplies the state value at runtime. Use when the route changes during operation (e.g. ADC source selection).
- mux-states = <&ctrl addressing-cells state-value> — the desired state is baked into devicetree. The consumer applies it once at init and never touches it again. Use when integration-time wiring is fixed (e.g. routing a QDEC's phase-A to a specific external pin)

Each controller declares #mux-control-cells / #mux-state-cells so the addressing layout matches the underlying hardware.

#### 2. Generic MUX driver API

Add a common MUX API in `include/zephyr/drivers/mux.h`

The API includes:

Function | Purpose
-- | --
mux_control_set(dev, ctrl, state) | Drive ctrl on dev to the given state. Used by mux-controls consumers.
mux_state_apply(dev, mstate) | Convenience for mux-states consumers — applies the DT-defined state in one call.
mux_state_get(dev, ctrl, *state) | Read back the current state. Optional; backends without read-back return -ENOSYS.
mux_lock(dev, ctrl, true) | Lock the control against further changes (e.g. TRGMUX one-way register lock). Optional.

#### 3. Controller drivers

Add the following MUX controller drivers:

- generic `gpio-mux`
- NXP `INPUTMUX`
- NXP `XBAR`
- NXP `TRGMUX`

These cover both simple generic GPIO-based multiplexers and SoC
routing peripherals used by NXP platforms.

#### 4. Driver integration

Convert existing drivers to use the MUX device model instead of controller-specific
handling:

- `qdec_mcux`
-  `qdc_mcux`

This allows these drivers to consume MUX routing information from devicetree
through the generic MUX API.

decsny commented:
I forgot to mark as draft when I made the PR, did not mean for the bot to ping all these people yet, sorry

decsny commented:
> The concept is sound, but it will definitely need a device driver agnostic framework to be feasible, specifically, a device driver, must not need to concern itself with whether its "on" a mux, similar to the way a device driver must not need to concern itself with whether its on a power domain:
> 
> ```
> &i2c1 {
> 	status = "okay";
> 
> 	sensor0: sensor@64 {
> 		...
> 	};
> };
> ```
> 
> must be identical to
> 
> ```
> i2c-mux {
> 	compatible = "i2c-mux";
> 	i2c-parent = <&i2c1>;
> 
> 	i2c@0 {
> 		reg = <0>;
> 		#address-cells = <1>;
> 		#size-cells = <0>;
> 		sensor0: sensor@3c {
> 			...
> 		};
> 	};
> };
> ```
> 
> from both the sensor device driver and application's perspective. Otherwise, every single device which could possibly be a consumer of a mux would need to `MUX_DT_DEFINE()` and `MUX_DT_APPLY()` within the driver.
> 
> We have `device_init()` and `device_deinit()` which we could hook into, try to init two devices which require two different states of the mux, and we can fail the init. Device runtime PM could theoretically be used as well, but this would make the mux framework dependent on device PM.

the mux device is referenced by a phandle from the consumer, not the parent, so the structure is agnostic. Just like how there is an "interrupts tree" or "clocks tree" within DT there would be a "mux tree" as well that has nothing to do with the bus hierarchy structure

and yes, every device driver that is enabled to be a consumer of a mux would have MUX_DT_DEFINE in it, just like how we have with PINCTRL_DT_DEFINE for devices which are enabled to mux external pins and the like, this is not a problem

bjarki-andreasen commented:
> > The concept is sound, but it will definitely need a device driver agnostic framework to be feasible, specifically, a device driver, must not need to concern itself with whether its "on" a mux, similar to the way a device driver must not need to concern itself with whether its on a power domain:
> > ```
> > &i2c1 {
> > 	status = "okay";
> > 
> > 	sensor0: sensor@64 {
> > 		...
> > 	};
> > };
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > must be identical to
> > ```
> > i2c-mux {
> > 	compatible = "i2c-mux";
> > 	i2c-parent = <&i2c1>;
> > 
> > 	i2c@0 {
> > 		reg = <0>;
> > 		#address-cells = <1>;
> > 		#size-cells = <0>;
> > 		sensor0: sensor@3c {
> > 			...
> > 		};
> > 	};
> > };
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > from both the sensor device driver and application's perspective. Otherwise, every single device which could possibly be a consumer of a mux would need to `MUX_DT_DEFINE()` and `MUX_DT_APPLY()` within the driver.
> > We have `device_init()` and `device_deinit()` which we could hook into, try to init two devices which require two different states of the mux, and we can fail the init. Device runtime PM could theoretically be used as well, but this would make the mux framework dependent on device PM.
> 
> the mux device is referenced by a phandle from the consumer, not the parent, so the structure is agnostic. Just like how there is an "interrupts tree" or "clocks tree" within DT there would be a "mux tree" as well that has nothing to do with the bus hierarchy structure
> 
> and yes, every device driver that is enabled to be a consumer of a mux would have MUX_DT_DEFINE in it, just like how we have with PINCTRL_DT_DEFINE for devices which are enabled to mux external pinctrl groups and the like, this is not a problem

it is a problem, if I write an i2c attached sensor driver, it is not necessarily, even not likely, to be a mux consumer. This is different from a PINCTRL which for a given device driver which requires it, it is mandatory, its not an optional, unlikely scenario to track. There are hundreds of sensor drivers, adding this "just in case" MUX consumer tracking to every driver is not an option IMO.

decsny commented:
> > > The concept is sound, but it will definitely need a device driver agnostic framework to be feasible, specifically, a device driver, must not need to concern itself with whether its "on" a mux, similar to the way a device driver must not need to concern itself with whether its on a power domain:
> > > ```
> > > &i2c1 {
> > > 	status = "okay";
> > > 
> > > 	sensor0: sensor@64 {
> > > 		...
> > > 	};
> > > };
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > must be identical to
> > > ```
> > > i2c-mux {
> > > 	compatible = "i2c-mux";
> > > 	i2c-parent = <&i2c1>;
> > > 
> > > 	i2c@0 {
> > > 		reg = <0>;
> > > 		#address-cells = <1>;
> > > 		#size-cells = <0>;
> > > 		sensor0: sensor@3c {
> > > 			...
> > > 		};
> > > 	};
> > > };
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > from both the sensor device driver and application's perspective. Otherwise, every single device which could possibly be a consumer of a mux would need to `MUX_DT_DEFINE()` and `MUX_DT_APPLY()` within the driver.
> > > We have `device_init()` and `device_deinit()` which we could hook into, try to init two devices which require two different states of the mux, and we can fail the init. Device runtime PM could theoretically be used as well, but this would make the mux framework dependent on device PM.
> > 
> > 
> > the mux device is referenced by a phandle from the consumer, not the parent, so the structure is agnostic. Just like how there is an "interrupts tree" or "clocks tree" within DT there would be a "mux tree" as well that has nothing to do with the bus hierarchy structure
> > and yes, every device driver that is enabled to be a consumer of a mux would have MUX_DT_DEFINE in it, just like how we have with PINCTRL_DT_DEFINE for devices which are enabled to mux external pinctrl groups and the like, this is not a problem
> 
> it is a problem, if I write an i2c attached sensor driver, it is not necessarily, even not likely, to be a mux consumer. This is different from a PINCTRL which for a given device driver which requires it, it is mandatory, its not an optional unlikely scenario to track. There are hundreds of sensor drivers, adding this "just in case" MUX consumer tracking to every driver is not an option IMO.

I am not understanding what the problem is, you don't have to add muxing to each driver, especially if it will never have a mux to configure.  For example with this i2c sensor, the mux consumer would likely be the i2c controller, not the sensor.  Because the controller is the one that might need to mux something to receive triggers or dma requests and the like from other things, not the sensor. The sensor just uses the i2c api, and I doubt it would ever need to know about a mux, you are right. Whereas the i2c controllers will probably often need to mux something to enable some type of functionality such as DMA-based transfers. So again I am not seeing any issue here.

bjarki-andreasen commented:
> > > > The concept is sound, but it will definitely need a device driver agnostic framework to be feasible, specifically, a device driver, must not need to concern itself with whether its "on" a mux, similar to the way a device driver must not need to concern itself with whether its on a power domain:
> > > > ```
> > > > &i2c1 {
> > > > 	status = "okay";
> > > > 
> > > > 	sensor0: sensor@64 {
> > > > 		...
> > > > 	};
> > > > };
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > must be identical to
> > > > ```
> > > > i2c-mux {
> > > > 	compatible = "i2c-mux";
> > > > 	i2c-parent = <&i2c1>;
> > > > 
> > > > 	i2c@0 {
> > > > 		reg = <0>;
> > > > 		#address-cells = <1>;
> > > > 		#size-cells = <0>;
> > > > 		sensor0: sensor@3c {
> > > > 			...
> > > > 		};
> > > > 	};
> > > > };
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > from both the sensor device driver and application's perspective. Otherwise, every single device which could possibly be a consumer of a mux would need to `MUX_DT_DEFINE()` and `MUX_DT_APPLY()` within the driver.
> > > > We have `device_init()` and `device_deinit()` which we could hook into, try to init two devices which require two different states of the mux, and we can fail the init. Device runtime PM could theoretically be used as well, but this would make the mux framework dependent on device PM.
> > > 
> > > 
> > > the mux device is referenced by a phandle from the consumer, not the parent, so the structure is agnostic. Just like how there is an "interrupts tree" or "clocks tree" within DT there would be a "mux tree" as well that has nothing to do with the bus hierarchy structure
> > > and yes, every device driver that is enabled to be a consumer of a mux would have MUX_DT_DEFINE in it, just like how we have with PINCTRL_DT_DEFINE for devices which are enabled to mux external pinctrl groups and the like, this is not a problem
> > 
> > it is a problem, if I write an i2c attached sensor driver, it is not necessarily, even not likely, to be a mux consumer. This is different from a PINCTRL which for a given device driver which requires it, it is mandatory, its not an optional unlikely scenario to track. There are hundreds of sensor drivers, adding this "just in case" MUX consumer tracking to every driver is not an option IMO.
> 
> I am not understanding what the problem is, you don't have to add muxing to each driver, especially if it will never have a mux to configure.  For example with this i2c sensor, the mux consumer would likely be the i2c controller, not the sensor.  Because the controller is the one that might need to mux something to receive triggers or dma requests and the like from other things, not the sensor. The sensor just uses the i2c api, and I doubt it would ever need to know about a mux, you are right. Whereas the i2c controllers will probably often need to mux something to enable some type of functionality such as DMA-based transfers. So again I am not seeing any issue here.

Looking at this PR in isolation, it is essentially pinctrl, but for any signal with a proper device driver API, and in isolation, that is fine. Looking at the linux binding however, it primarily describes a separate use case, the usecase i summarized above, that is not being addressed in this PR is it?

bjarki-andreasen commented:
> > > > > The concept is sound, but it will definitely need a device driver agnostic framework to be feasible, specifically, a device driver, must not need to concern itself with whether its "on" a mux, similar to the way a device driver must not need to concern itself with whether its on a power domain:
> > > > > ```
> > > > > &i2c1 {
> > > > > 	status = "okay";
> > > > > 
> > > > > 	sensor0: sensor@64 {
> > > > > 		...
> > > > > 	};
> > > > > };
> > > > > ```
> > > > > 
> > > > > 
> > > > >     
> > > > >       
> > > > >     
> > > > > 
> > > > >       
> > > > >     
> > > > > 
> > > > >     
> > > > >   
> > > > > must be identical to
> > > > > ```
> > > > > i2c-mux {
> > > > > 	compatible = "i2c-mux";
> > > > > 	i2c-parent = <&i2c1>;
> > > > > 
> > > > > 	i2c@0 {
> > > > > 		reg = <0>;
> > > > > 		#address-cells = <1>;
> > > > > 		#size-cells = <0>;
> > > > > 		sensor0: sensor@3c {
> > > > > 			...
> > > > > 		};
> > > > > 	};
> > > > > };
> > > > > ```
> > > > > 
> > > > > 
> > > > >     
> > > > >       
> > > > >     
> > > > > 
> > > > >       
> > > > >     
> > > > > 
> > > > >     
> > > > >   
> > > > > from both the sensor device driver and application's perspective. Otherwise, every single device which could possibly be a consumer of a mux would need to `MUX_DT_DEFINE()` and `MUX_DT_APPLY()` within the driver.
> > > > > We have `device_init()` and `device_deinit()` which we could hook into, try to init two devices which require two different states of the mux, and we can fail the init. Device runtime PM could theoretically be used as well, but this would make the mux framework dependent on device PM.
> > > > 
> > > > 
> > > > the mux device is referenced by a phandle from the consumer, not the parent, so the structure is agnostic. Just like how there is an "interrupts tree" or "clocks tree" within DT there would be a "mux tree" as well that has nothing to do with the bus hierarchy structure
> > > > and yes, every device driver that is enabled to be a consumer of a mux would have MUX_DT_DEFINE in it, just like how we have with PINCTRL_DT_DEFINE for devices which are enabled to mux external pinctrl groups and the like, this is not a problem
> > > 
> > > it is a problem, if I write an i2c attached sensor driver, it is not necessarily, even not likely, to be a mux consumer. This is different from a PINCTRL which for a given device driver which requires it, it is mandatory, its not an optional unlikely scenario to track. There are hundreds of sensor drivers, adding this "just in case" MUX consumer tracking to every driver is not an option IMO.
> > 
> > I am not understanding what the problem is, you don't have to add muxing to each driver, especially if it will never have a mux to configure.  For example with this i2c sensor, the mux consumer would likely be the i2c controller, not the sensor.  Because the controller is the one that might need to mux something to receive triggers or dma requests and the like from other things, not the sensor. The sensor just uses the i2c api, and I doubt it would ever need to know about a mux, you are right. Whereas the i2c controllers will probably often need to mux something to enable some type of functionality such as DMA-based transfers. So again I am not seeing any issue here.
> 
> Looking at this PR in isolation, it is essentially pinctrl, but for any signal with a proper device driver API, and in isolation, that is fine. Looking at the linux binding however, it primarily describes a separate use case, the usecase i summarized above, that is not being addressed in this PR is it? In the above, neither the i2c controller, nor the sensor, care at all whether there is a mux in between them

decsny commented:
@carlescufi @cfriedt 

I checked on what this devmux thing is, and I came to the conclusion that it is completely unrelated.  What I am adding here is a driver API class for a type of hardware peripheral that functions as a multiplexer.  Devmux is mostly looking like a software trick that allows one "virtual device" to be routed to multiple different ones, using the devmux API, which under the hood essentially dynamically overwrites the contents of it's own `struct device` to match a different device.  It looks like the only example use case in tree is routing "printf" to different uarts at runtime.  Which is a neat thing, kudos to Chris for that idea, but indeed is not remotely related to this RFC I would say.

bjarki-andreasen reviewed (dismissed):
Design is nice. some nits though

decsny commented:
I might continue working on this but I will probably fork it and remove some of the commits as I no longer have access to the source branch repo or the specific hardware I was using before. So I am going to close this ticket and I will develop it for a different integration platform if I continue working on this.

decsny commented:
I intend to push for this design still but have lost access to the source branch. @FelixWang47831  expressed interest to help with this so I am reopening this PR.  I might move it to a new PR so I can work on it.  But I know the NXP folks are required to work off of this repo for this PR so collaboration might awkwardly require both PR to be open

FelixWang47831 commented on include/zephyr/drivers/mux.h:
```diff
@@ -0,0 +1,445 @@
+				       uint32_t *state)
+{
+	const struct mux_control_driver_api *api =
+		(const struct mux_control_driver_api *)dev->api;
```
Updated.

pdgendt commented on drivers/mux/mux_reg_mux.c:
Ditto for 8100f92e7888327fb742cbd8812733d79b3c62b1 and a774cc69343e87d2180b4535fe8d5db1e66a40e0

pdgendt reviewed (commented):
Some of the commits update drivers that are added by this PR, I think these should be squashed/cleanup up.

Commit:
drivers: mux: introduce mux controller subsystem

Introduce a new generic multiplexer (MUX) controller subsystem
modeled on the Linux mux framework. Provides a unified API for
software- and hardware-level signal routers via mux_control_set()
and mux_state_apply(), with userspace syscall handlers and common
devicetree bindings (mux-controller / mux-consumer) for downstream
consumers.

Co-Authored-By: Felix Wang <fei.wang_3@nxp.com>
Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
drivers: mux: add GPIO-based mux driver

Add a generic GPIO-backed mux controller implementation: the
selected state value is binary-encoded LSB-first across the GPIO
lines listed in the mux-gpios property. Useful for board-level
analog/digital muxes that are simply wired to MCU GPIOs.

Co-Authored-By: Felix Wang <fei.wang_3@nxp.com>
Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

FelixWang47831 committed:
drivers: mux: add NXP INPUTMUX mux driver

Add a mux controller driver for the NXP INPUTMUX peripheral.
Wraps the HAL INPUTMUX_AttachSignal() routine to route a
peripheral input signal onto a destination channel through the
unified mux_state_apply() API.

Co-Authored-By: Declan Snyder <declan.snyder@nxp.com>
Signed-off-by: Felix Wang <fei.wang_3@nxp.com>

FelixWang47831 committed:
drivers: mux: add NXP TRGMUX mux driver

Add a mux controller driver for the NXP TRGMUX peripheral, used to
multiplex trigger signals between on-chip peripherals. The
controller exposes two address cells (trgmux instance / output id)
and supports register-level lock so that an applied routing can be
frozen for the lifetime of the system.

Signed-off-by: Felix Wang <fei.wang_3@nxp.com>

FelixWang47831 committed:
drivers: clock_control: imx_ccm: add XBAR clock support

Define the IMX_CCM_XBAR1_CLK and IMX_CCM_XBAR2_CLK identifiers and
wire them into the i.MX CCM clock_control driver so that the XBAR
mux driver can gate its peripheral clock on/off. Guarded by
CONFIG_MUX_NXP_XBAR so the code is only built when the XBAR mux
driver is selected.

Signed-off-by: Felix Wang <fei.wang_3@nxp.com>

FelixWang47831 committed:
drivers: mux: add NXP MCUX XBAR mux driver

Add a mux controller driver for the NXP MCUX XBAR (Crossbar
Switch) peripheral, supporting both 16-bit and 32-bit SEL register
layouts via DEVICE_MMIO. Update the nxp,mcux-xbar binding to also
declare the mux-controller interface (#mux-control-cells and
#mux-state-cells) and require the CCM clock phandle, while still
documenting the legacy xbar-maps property for as-yet-unmigrated
consumers (e.g. qdec_mcux).

Signed-off-by: Felix Wang <fei.wang_3@nxp.com>

FelixWang47831 committed:
drivers: sensor: nxp: qdc_mcux: migrate to mux subsystem

Drop the direct INPUTMUX_AttachSignal() coupling in the QDC driver
in favour of the new generic mux subsystem: the driver now selects
CONFIG_MUX, defines mux states from a "mux-states" devicetree
property, and applies them at init time via mux_state_apply().
This makes the driver mux-hardware-agnostic.

Update the nxp,mcux-qdc binding to use mux-consumer.yaml, expose
the inputmux0 node on the MCXNx4x family DTSI, and migrate the
FRDM-MCXN947 / MCXN5xx-EVK QDC overlays to the new mux-states
syntax.

Signed-off-by: Felix Wang <fei.wang_3@nxp.com>

FelixWang47831 committed:
drivers: sensor: nxp: qdec_mcux: migrate to mux subsystem

Drop the direct MCUX_XBARA HAL coupling in the QDEC driver in
favour of the new generic mux subsystem: the driver now selects
CONFIG_MUX, defines mux states from a "mux-states" devicetree
property and applies them at init time via mux_state_apply(),
which makes it agnostic to the underlying mux hardware.

Update the nxp,mcux-qdec binding to include mux-consumer.yaml,
augment the i.MX RT10xx XBAR nodes with CCM clock phandles plus
the new #mux-control-cells / #mux-state-cells declarations, and
migrate the MIMXRT1050-EVK QDEC overlay to the mux-states syntax.

Signed-off-by: Felix Wang <fei.wang_3@nxp.com>

FelixWang47831 commented:
The commits have been reorganized to make them more concise.
Update:
1. Previous design uses one `struct mux_control` + `enum mux_spec_type` to indicate the data is a mux_control or mux_state and reports type mismatches at runtime (-ESRCH from `mux_configure_default`).
      
    Now update to use two distinct types: `struct mux_control` (addressing
      only) and `struct mux_state` (addressing + default state value).
      Misuse is rejected at compile time; no runtime spec-type tag is needed.

2. API name changed:
  a. Configure a MUX control to a specific state
          mux_configure(const struct device *dev, struct mux_control *control, uint32_t state);
          change to:
          mux_control_set(const struct device *dev, const struct mux_control *control, uint32_t state);

    b. Configure a MUX control to its default state from devicetree
           mux_configure_default(const struct device *dev, struct mux_control *control)
           change to:
           mux_state_apply(const struct device *dev, const struct mux_state *mstate)

I made this change because I think it's unclear to mix mux control and mux state structure, and I find Linux driver
also distinguish these two type data. Linux use API mux_control_select, mux_state_select, I think it's more clearly indicates the target of the operation, so I update API name.
see https://github.com/torvalds/linux/blob/master/include/linux/mux/consumer.h

FelixWang47831 commented:
Hi @pdgendt @bjarki-andreasen @mbolivar @cfriedt 
Thank you for your previous review. Regarding the current commit, do you have any further comments? I would like to continue moving this PR forward. Thank you.

FelixWang47831 commented:
Since the discussion history is too long and no one is looking at it anymore, and I'm not sure about its architecture review status because @decsny has left the company, I just create a new PR https://github.com/zephyrproject-rtos/zephyr/pull/112088. If it needs to be reviewed at the WG meeting, I can explain it again then.
